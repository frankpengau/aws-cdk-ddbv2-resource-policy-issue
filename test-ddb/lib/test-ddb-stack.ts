import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import * as iam from 'aws-cdk-lib/aws-iam';

export class TestDdbStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const roles = ['test-ddb-role-1', 'test-ddb-role-2', 'test-ddb-role-3'];

    roles.forEach(roleName => {
      new iam.Role(this, roleName, {
        roleName: roleName,
        assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com'),
        inlinePolicies: {
          DynamoDBAccess: new iam.PolicyDocument({
            statements: [
              new iam.PolicyStatement({
                actions: [
                  'dynamodb:*'
                ],
                effect: iam.Effect.ALLOW,
                resources: [
                  "arn:aws:dynamodb:ap-southeast-2:123456789012:table/ddb-table-test",
                  "arn:aws:dynamodb:ap-southeast-2:123456789012:table/ddb-table-test/index/*",
                ]
              })
            ]
          })
        }
      });
    });

    const table = new dynamodb.TableV2(this, 'Table', {
      tableName: 'ddb-table-test',
      partitionKey: { name: 'id', type: dynamodb.AttributeType.STRING },
      resourcePolicy: new iam.PolicyDocument({
        statements: [
          new iam.PolicyStatement({
            actions: [
              'dynamodb:GetShardIterator',
              'dynamodb:Scan',
              'dynamodb:Query',
              'dynamodb:DescribeStream',
              'dynamodb:GetRecords',
              'dynamodb:ListStreams',
            ],
            effect: iam.Effect.ALLOW,
            principals: [
              new iam.ArnPrincipal('arn:aws:iam::123456789012:role/test-ddb-role-1'),
              new iam.ArnPrincipal('arn:aws:iam::123456789012:role/test-ddb-role-2'),
              new iam.ArnPrincipal('arn:aws:iam::123456789012:role/test-ddb-role-3')
            ],
            resources: [
              "arn:aws:dynamodb:ap-southeast-2:123456789012:table/ddb-table-test/index/*",
              "arn:aws:dynamodb:ap-southeast-2:123456789012:table/ddb-table-test/stream/*"
            ],
          }),
          new iam.PolicyStatement({
            actions: [
              "dynamodb:BatchGetItem",
              "dynamodb:BatchWriteItem",
              "dynamodb:ConditionCheckItem",
              "dynamodb:PutItem",
              "dynamodb:DescribeTable",
              "dynamodb:DeleteItem",
              "dynamodb:GetItem",
              "dynamodb:Scan",
              "dynamodb:Query",
              "dynamodb:UpdateItem"
            ],
            effect: iam.Effect.ALLOW,
            principals: [
              new iam.ArnPrincipal('arn:aws:iam::123456789012:role/test-ddb-role-1'),
              new iam.ArnPrincipal('arn:aws:iam::123456789012:role/test-ddb-role-2'),
              new iam.ArnPrincipal('arn:aws:iam::123456789012:role/test-ddb-role-3')
            ],
            resources: [
              "arn:aws:dynamodb:ap-southeast-2:123456789012:table/ddb-table-test"
            ],
          })
        ]
      })
    });
  }
}
