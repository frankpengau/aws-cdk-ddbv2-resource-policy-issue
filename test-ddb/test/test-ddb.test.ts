import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { TestDdbStack } from '../lib/test-ddb-stack';

test('DynamoDB Stack Snapshot', () => {
  const app = new cdk.App();
  const stack = new TestDdbStack(app, 'TestDdbStack');

  const template = Template.fromStack(stack);

  expect(template.toJSON()).toMatchSnapshot();
});