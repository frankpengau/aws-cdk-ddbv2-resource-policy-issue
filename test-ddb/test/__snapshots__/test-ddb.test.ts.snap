// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamoDB Stack Snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "TableCD117FA1": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "id",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
        ],
        "Replicas": [
          {
            "Region": {
              "Ref": "AWS::Region",
            },
            "ResourcePolicy": {
              "PolicyDocument": {
                "Statement": [
                  {
                    "Action": [
                      "dynamodb:GetShardIterator",
                      "dynamodb:Scan",
                      "dynamodb:Query",
                      "dynamodb:DescribeStream",
                      "dynamodb:GetRecords",
                      "dynamodb:ListStreams",
                    ],
                    "Effect": "Allow",
                    "Principal": {
                      "AWS": [
                        "arn:aws:iam::123456789012:role/test-ddb-role-1",
                        "arn:aws:iam::123456789012:role/test-ddb-role-2",
                        "arn:aws:iam::123456789012:role/test-ddb-role-3",
                      ],
                    },
                    "Resource": [
                      "arn:aws:dynamodb:ap-southeast-2:123456789012:table/ddb-table-test/index/*",
                      "arn:aws:dynamodb:ap-southeast-2:123456789012:table/ddb-table-test/stream/*",
                    ],
                  },
                  {
                    "Action": [
                      "dynamodb:BatchGetItem",
                      "dynamodb:BatchWriteItem",
                      "dynamodb:ConditionCheckItem",
                      "dynamodb:PutItem",
                      "dynamodb:DescribeTable",
                      "dynamodb:DeleteItem",
                      "dynamodb:GetItem",
                      "dynamodb:Scan",
                      "dynamodb:Query",
                      "dynamodb:UpdateItem",
                    ],
                    "Effect": "Allow",
                    "Principal": {
                      "AWS": [
                        "arn:aws:iam::123456789012:role/test-ddb-role-1",
                        "arn:aws:iam::123456789012:role/test-ddb-role-2",
                        "arn:aws:iam::123456789012:role/test-ddb-role-3",
                      ],
                    },
                    "Resource": "arn:aws:dynamodb:ap-southeast-2:123456789012:table/ddb-table-test",
                  },
                ],
                "Version": "2012-10-17",
              },
            },
          },
        ],
        "TableName": "ddb-table-test",
      },
      "Type": "AWS::DynamoDB::GlobalTable",
      "UpdateReplacePolicy": "Retain",
    },
    "testddbrole1EE1FCF5C": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "dynamodb:*",
                  "Effect": "Allow",
                  "Resource": [
                    "arn:aws:dynamodb:ap-southeast-2:123456789012:table/ddb-table-test",
                    "arn:aws:dynamodb:ap-southeast-2:123456789012:table/ddb-table-test/index/*",
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "DynamoDBAccess",
          },
        ],
        "RoleName": "test-ddb-role-1",
      },
      "Type": "AWS::IAM::Role",
    },
    "testddbrole2C5F6B0E6": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "dynamodb:*",
                  "Effect": "Allow",
                  "Resource": [
                    "arn:aws:dynamodb:ap-southeast-2:123456789012:table/ddb-table-test",
                    "arn:aws:dynamodb:ap-southeast-2:123456789012:table/ddb-table-test/index/*",
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "DynamoDBAccess",
          },
        ],
        "RoleName": "test-ddb-role-2",
      },
      "Type": "AWS::IAM::Role",
    },
    "testddbrole34961B6CD": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ec2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": "dynamodb:*",
                  "Effect": "Allow",
                  "Resource": [
                    "arn:aws:dynamodb:ap-southeast-2:123456789012:table/ddb-table-test",
                    "arn:aws:dynamodb:ap-southeast-2:123456789012:table/ddb-table-test/index/*",
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "DynamoDBAccess",
          },
        ],
        "RoleName": "test-ddb-role-3",
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
